### ISSUE TITLE
bug: Vocone can not download from IPFS censuses created by census3 

### ISSUE THREAD

**Issue by lucasmenendez** [vocdoni/vocdoni-node#1126]
### Describe the bug

Using the [deploy-snippets](https://github.com/vocdoni/deploy-snippets) repo I found an error during the creation of elections using a *"Token-based census"* from Census3.

Census3 syncs, creates and publishes successfully the census to IPFS but the Vocone can not download it, so the census never is created and the voters can cast any vote.

To validate the published census, you can use the following [gist](https://gist.github.com/lucasmenendez/36fc09785c9efbc60e92d51b2c84a3ce). Example:
```sh
go run main.go -uri ifps://...
```

### To Reproduce (please complete the following information)

- Run the whole stack using [deploy-snippets](https://github.com/vocdoni/deploy-snippets). Be sure that you defines a valid goerli web3 endpoint for census3.
- Create a test token:
```sh
curl -X POST \
  'http://localhost:8083/api/token' \
  --header 'Accept: */*' \
  --header 'Content-Type: application/json' \
  --data-raw '{
    "id": "0xf530280176385af31177d78bbfd5ea3f6d07488a",
    "type": "erc20",
    "tag": "test2",
    "chainID": 5
}'
```
- Wait to be synced. Check it with the following `curl`:
```sh
curl  -X GET \
  'http://localhost:8083/api/token/0xf530280176385af31177d78bbfd5ea3f6d07488a' \
  --header 'Accept: */*'
```
- The go to `localhost:8080` and create an election with a token-based census using the `Token Monkeys (MON)`.
- Look at the vocone logs.

**Comment by altergui** [comment]
so, upon investigation, this is due to a bug in `voconed`, that doesn't init the offchaindatadownloader properly
it just does:

https://github.com/vocdoni/vocdoni-node/blob/ddea83d2a0fd2b463475d00be1d37ffaac9beb3e/vocone/vocone.go#L129-L136

crucially, it's missing `vs.DataDownloader.Start()` that `node` does during init

https://github.com/vocdoni/vocdoni-node/blob/61101a8e4f32103c24a8e28fee2d6715fd9a7af6/service/offchaindata.go#L16-L37

this is a simple fix

```diff
diff --git a/vocone/vocone.go b/vocone/vocone.go
index 568e0444..6200ea99 100644
--- a/vocone/vocone.go
+++ b/vocone/vocone.go
@@ -129,12 +129,12 @@ func NewVocone(dataDir string, keymanager *ethereum.SignKeys, disableIPFS bool)
 
        // Create the data downloader and offchain data handler
        if !disableIPFS {
-               offchaindatahandler.NewOffChainDataHandler(
-                       vc.app,
-                       downloader.NewDownloader(vc.storage),
-                       vc.censusdb,
-                       false,
-               )
+               srv.App = vc.app
+               srv.Storage = vc.storage
+               srv.CensusDB = vc.censusdb
+               globalCfg := config.NewConfig()
+               srv.Config = globalCfg.Vochain
+               srv.OffChainDataHandler()
        }
 
        return vc, err
```

but i'll try refactoring the whole voconed, to make it much more similar to vocdoni-node, to diminish these kind of bugs and ease future maintenance

**Comment by altergui** [comment]
WORKSFORME

```diff
diff --git a/docker-compose.yml b/docker-compose.yml
index e964703..7ee1689 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -29,6 +29,8 @@ services:
   voconed:
     # pull main tag (instead of latest) to ensure we're testing against bleeding edge
     image: vocdoni/vocdoni-node:main
+    build:
+      context: https://github.com/vocdoni/vocdoni-node.git#f/offchaindh-fixes
     entrypoint: /app/voconed
     environment:
       - VOCONED_PORT=9095
```

after building the voconed docker image from that branch, i see the census being downloaded and added to censusdb

```
vocdoni-demo-voconed-1  | 2023-10-03T10:34:29.546Z DBG middleware/logger.go:158 > "GET http://localhost:8081/v2/elections/98591f8979815b106c84de3299489807e3b3ce39bb6c7c279d97020600000000 HTTP/1.1" from 172.21.0.1:50342 - 200 1716B in 789.276µs
vocdoni-demo-voconed-1  | 2023-10-03T10:34:29.574Z DBG httprouter/message.go:65 > response: (404) {"error":"census not found","code":4046}
vocdoni-demo-voconed-1  | 2023-10-03T10:34:29.574Z DBG middleware/logger.go:158 > "GET http://localhost:8081/v2/censuses/778ee1ac80aef0b79563228364570a3f9a55077c9bd13dcb7f5ddc386bd1461e/size HTTP/1.1" from 172.21.0.1:50342 - 404 41B in 272.47µs
vocdoni-demo-voconed-1  | 2023-10-03T10:34:29.575Z DBG httprouter/message.go:65 > response: (404) {"error":"census not found","code":4046}
vocdoni-demo-voconed-1  | 2023-10-03T10:34:29.575Z DBG middleware/logger.go:158 > "GET http://localhost:8081/v2/censuses/778ee1ac80aef0b79563228364570a3f9a55077c9bd13dcb7f5ddc386bd1461e/type HTTP/1.1" from 172.21.0.1:50348 - 404 41B in 213.282µs
vocdoni-demo-voconed-1  | 2023-10-03T10:34:29.575Z DBG httprouter/message.go:65 > response: (404) {"error":"census not found","code":4046}
vocdoni-demo-voconed-1  | 2023-10-03T10:34:29.575Z DBG middleware/logger.go:158 > "GET http://localhost:8081/v2/censuses/778ee1ac80aef0b79563228364570a3f9a55077c9bd13dcb7f5ddc386bd1461e/weight HTTP/1.1" from 172.21.0.1:50352 - 404 41B in 238.862µs
vocdoni-demo-voconed-1  | 2023-10-03T10:34:30.886Z DBG vocone/vocone.go:194 > block committed hash=0a4ab0b1eb91ed7334510e3df9aba51534c5ecc4fb034b39958bd280315afe6a height=1166 took=16.649271
vocdoni-demo-voconed-1  | 2023-10-03T10:34:35.842Z INF vocone/vocone.go:496 > [vochain info] height:1166 mempool:0 processes:1 votes:0 vxm:0 voteCache:0 blockTime:{1m:5.00}
vocdoni-demo-voconed-1  | 2023-10-03T10:34:35.887Z DBG vocone/vocone.go:194 > block committed hash=0a4ab0b1eb91ed7334510e3df9aba51534c5ecc4fb034b39958bd280315afe6a height=1167 took=16.798774
vocdoni-demo-voconed-1  | 2023-10-03T10:34:40.879Z DBG vocone/vocone.go:194 > block committed hash=0a4ab0b1eb91ed7334510e3df9aba51534c5ecc4fb034b39958bd280315afe6a height=1168 took=9.723438
vocdoni-demo-voconed-1  | 2023-10-03T10:34:45.843Z INF vocone/vocone.go:496 > [vochain info] height:1168 mempool:0 processes:1 votes:0 vxm:0 voteCache:0 blockTime:{1m:5.00}
vocdoni-demo-voconed-1  | 2023-10-03T10:34:45.879Z DBG vocone/vocone.go:194 > block committed hash=0a4ab0b1eb91ed7334510e3df9aba51534c5ecc4fb034b39958bd280315afe6a height=1169 took=8.377227
vocdoni-demo-voconed-1  | 2023-10-03T10:34:50.29Z DBG ipfs/ipfs.go:319 > rawdata received: {"type":2,"rootHash":"d47hrICu8LeVYyKDZFcKP5pVB3yb0T3Lf13cOGvRRh4=","data":"KLUv/UQAXQHNCgCEExQgABiT7XhIAmfRhUNzqZzujeLgjUMPAADITmdtwRsAFCAAONK8kbiZKPeMurPksZSeKHh+x6MAABSAH0dptAAAFCAAKkY2paETjjX3+T6B+lbTyXC8Z3cAAGSns7bgDQAUIADeuGmWWb5dQaDlfhedbLQuALkgDAAAhOJQbOZ8ABQgAOEwio0CkYSb/7IAvlgstjR/vpDZMTx/cSZIb/78qp/qktloy/iRuAwAACz2GiSiKbHwWxG6PYku3QDy52iXeeK4hBgnAAD0RIKRY0VF06A+gwLeZZ5+p0AMTP6crtjHIwAAWOw1SERT5U1wL5jjEqukMY48a9upirXhEBIAAEB2OmsL3vPEVvqqcP6jB6Bzw9qVckE8d/WL91K1J+Kro5XRukwN6cFHt2PdofQAALDYa5CIpgsAAOiIA8kAoQOEDhA6QOigFTkAapKsajJWSga1IwPdVRLm","maxLevels":160}
vocdoni-demo-voconed-1  | 2023-10-03T10:34:50.291Z INF ipfs/ipfs.go:325 > retrieved file path=/ipfs/bafybeiezzvgxwv6nykjkbcmqfd55g7smpilvj7nedx2crynyhtp2u2tx5a size=574
vocdoni-demo-voconed-1  | 2023-10-03T10:34:50.291Z DBG censusdb/censusdb.go:181 > importing census root=778ee1ac80aef0b79563228364570a3f9a55077c9bd13dcb7f5ddc386bd1461e type=ARBO_POSEIDON
vocdoni-demo-voconed-1  | 2023-10-03T10:34:50.323Z DBG compressor/compression.go:78 > decompressed 0.36 KiB to 0.60 KiB in 31.884µs with zstd, 168.1% of the original size
vocdoni-demo-voconed-1  | 2023-10-03T10:34:50.887Z DBG vocone/vocone.go:194 > block committed hash=0a4ab0b1eb91ed7334510e3df9aba51534c5ecc4fb034b39958bd280315afe6a height=1170 took=16.266203
vocdoni-demo-voconed-1  | 2023-10-03T10:34:55.764Z INF offchain downloader enqueued=0 retrying=0 total=4
vocdoni-demo-voconed-1  | 2023-10-03T10:34:55.844Z INF vocone/vocone.go:496 > [vochain info] height:1170 mempool:0 processes:1 votes:0 vxm:0 voteCache:0 blockTime:{1m:5.00}
vocdoni-demo-voconed-1  | 2023-10-03T10:34:55.879Z DBG vocone/vocone.go:194 > block committed hash=0a4ab0b1eb91ed7334510e3df9aba51534c5ecc4fb034b39958bd280315afe6a height=1171 took=8.054474
vocdoni-demo-voconed-1  | 2023-10-03T10:34:55.901Z INF ipfs storage addresses=3218 peers=93 pins=3
vocdoni-demo-voconed-1  | 2023-10-03T10:35:00.889Z DBG vocone/vocone.go:194 > block committed hash=0a4ab0b1eb91ed7334510e3df9aba51534c5ecc4fb034b39958bd280315afe6a height=1172 took=16.567377
vocdoni-demo-voconed-1  | 2023-10-03T10:35:00.998Z DBG ipfs/ipfs.go:271 > retrieved file /ipfs/bafybeialqkqfsyocfh6oi4ngkvvna7xognsfbvfwo57qjbsxlrqxo2zpjm from cache
vocdoni-demo-voconed-1  | 2023-10-03T10:35:00.998Z DBG httprouter/message.go:65 > response: (200) {"electionId":"98591f8979815b106c84de3299489807e3b3ce39bb6c7c279d97020600000000","organizationId":"5b106c84de3299489807e3b3ce39bb6c7c279d97","status":"READY","startDate":"0001-01-01T00:00:00Z","endDate":"2023-10-04T08:10:20.99804572Z","voteCount":0,"finalResults":false,"manuallyEnded":false,"census":{"censusOrigin":"OFF_CHAIN_TREE_WEIGHTED","censusRoot":"778ee1ac80aef0b79563228364570a3f9a55077c9bd13dcb7f5ddc386bd1461e","postRegisterCensusRoot":"","censusURL":"ipfs://bafybeiezzvgxwv6nykjkbcmqfd55g7smpilvj7nedx2crynyhtp2u2tx5a","maxCensusSize":6},"metadataURL":"ipfs://bafybeialqkqfsyocfh6oi4ngkvvna7xognsfbvfwo57qjbsxlrqxo2zpjm","creationTime":"2023-10-03T10:33:45Z","voteMode":{"serial":false,"anonymous":true,"encryptedVotes":true,"uniqueValues":false,"costFromWeight":false},"electionMode":{"autoStart":true,"interruptible":true,"dynamicCensus":false,"encryptedMetaData":false,"preRegister":false},"tallyMode":{"maxCount":1,"maxValue":1,"maxVoteOverwrites":1,"maxTotalCost":0,"costExponent":10000},"metadata":{"title":{"default":"fdsa"},"version":"1.1","description":{"default":""},"media":{},"meta":{"census":{"type":"token"},"generated":"ui-scaffold","sdk":{"version":"0.3.1"},"token":{"chainID":5,"decimals":18,"defaultStrategy":1,"id":"0xF530280176385AF31177D78BbFD5eA3f6D07488A","name":"Monkeys","size":11,"startBlock":8901659,"status":{"atBlock":9801601,"progress":100,"synced":true},"symbol":"MON","tag":"test2","totalSupply":"82000000000000000000","type":"erc20"}},"questions":[{"choices":[{"title":{"default":"1"},"value":0},{"title":{"default":"1"},"value":1}],"description":{"default":""},"title":{"default":"fdsa"}}],"results":{"aggregation":"discrete-counting","display":"multiple-question"}}}
vocdoni-demo-voconed-1  | 2023-10-03T10:35:00.998Z DBG middleware/logger.go:158 > "GET http://localhost:8081/v2/elections/98591f8979815b106c84de3299489807e3b3ce39bb6c7c279d97020600000000 HTTP/1.1" from 172.21.0.1:41824 - 200 1715B in 599.4µs
vocdoni-demo-voconed-1  | 2023-10-03T10:35:01.029Z DBG censusdb/censusdb.go:134 > loaded tree 778ee1ac80aef0b79563228364570a3f9a55077c9bd13dcb7f5ddc386bd1461e of type ARBO_POSEIDON
vocdoni-demo-voconed-1  | 2023-10-03T10:35:01.029Z DBG censusdb/censusdb.go:134 > loaded tree 778ee1ac80aef0b79563228364570a3f9a55077c9bd13dcb7f5ddc386bd1461e of type ARBO_POSEIDON
vocdoni-demo-voconed-1  | 2023-10-03T10:35:01.029Z DBG censusdb/censusdb.go:134 > loaded tree 778ee1ac80aef0b79563228364570a3f9a55077c9bd13dcb7f5ddc386bd1461e of type ARBO_POSEIDON
vocdoni-demo-voconed-1  | 2023-10-03T10:35:01.029Z DBG httprouter/message.go:65 > response: (200) {"size":11}
vocdoni-demo-voconed-1  | 2023-10-03T10:35:01.029Z DBG httprouter/message.go:65 > response: (200) {"weight":"82000000000000000000"}
vocdoni-demo-voconed-1  | 2023-10-03T10:35:01.029Z DBG middleware/logger.go:158 > "GET http://localhost:8081/v2/censuses/778ee1ac80aef0b79563228364570a3f9a55077c9bd13dcb7f5ddc386bd1461e/weight HTTP/1.1" from 172.21.0.1:41842 - 200 34B in 8.804763ms
vocdoni-demo-voconed-1  | 2023-10-03T10:35:01.029Z DBG httprouter/message.go:65 > response: (200) {"type":"zkweighted"}
vocdoni-demo-voconed-1  | 2023-10-03T10:35:01.029Z DBG middleware/logger.go:158 > "GET http://localhost:8081/v2/censuses/778ee1ac80aef0b79563228364570a3f9a55077c9bd13dcb7f5ddc386bd1461e/type HTTP/1.1" from 172.21.0.1:41830 - 200 22B in 8.977982ms
vocdoni-demo-voconed-1  | 2023-10-03T10:35:01.029Z DBG middleware/logger.go:158 > "GET http://localhost:8081/v2/censuses/778ee1ac80aef0b79563228364570a3f9a55077c9bd13dcb7f5ddc386bd1461e/size HTTP/1.1" from 172.21.0.1:41824 - 200 12B in 9.535428ms
vocdoni-demo-voconed-1  | 2023-10-03T10:35:05.845Z INF vocone/vocone.go:496 > [vochain info] height:1172 mempool:0 processes:1 votes:0 vxm:0 voteCache:0 blockTime:{1m:5.00}
vocdoni-demo-voconed-1  | 2023-10-03T10:35:05.889Z DBG vocone/vocone.go:194 > block committed hash=0a4ab0b1eb91ed7334510e3df9aba51534c5ecc4fb034b39958bd280315afe6a height=1173 took=16.642322
vocdoni-demo-voconed-1  | 2023-10-03T10:35:10.89Z DBG vocone/vocone.go:194 > block committed hash=0a4ab0b1eb91ed7334510e3df9aba51534c5ecc4fb034b39958bd280315afe6a height=1174 took=16.462885
vocdoni-demo-voconed-1  | 2023-10-03T10:35:15.846Z INF vocone/vocone.go:496 > [vochain info] height:1174 mempool:0 processes:1 votes:0 vxm:0 voteCache:0 blockTime:{1m:5.00}
vocdoni-demo-voconed-1  | 2023-10-03T10:35:15.891Z DBG vocone/vocone.go:194 > block committed hash=0a4ab0b1eb91ed7334510e3df9aba51534c5ecc4fb034b39958bd280315afe6a height=1175 took=16.499367
vocdoni-demo-voconed-1  | 2023-10-03T10:35:20.892Z DBG vocone/vocone.go:194 > block committed hash=0a4ab0b1eb91ed7334510e3df9aba51534c5ecc4fb034b39958bd280315afe6a height=1176 took=16.528727
vocdoni-demo-voconed-1  | 2023-10-03T10:35:25.847Z INF vocone/vocone.go:496 > [vochain info] height:1176 mempool:0 processes:1 votes:0 vxm:0 voteCache:0 blockTime:{1m:5.00}
vocdoni-demo-voconed-1  | 2023-10-03T10:35:25.892Z DBG vocone/vocone.go:194 > block committed hash=0a4ab0b1eb91ed7334510e3df9aba51534c5ecc4fb034b39958bd280315afe6a height=1177 took=16.389069
vocdoni-demo-voconed-1  | 2023-10-03T10:35:30.893Z DBG vocone/vocone.go:194 > block committed hash=0a4ab0b1eb91ed7334510e3df9aba51534c5ecc4fb034b39958bd280315afe6a height=1178 took=16.941371
vocdoni-demo-voconed-1  | 2023-10-03T10:35:31.067Z DBG ipfs/ipfs.go:271 > retrieved file /ipfs/bafybeialqkqfsyocfh6oi4ngkvvna7xognsfbvfwo57qjbsxlrqxo2zpjm from cache
vocdoni-demo-voconed-1  | 2023-10-03T10:35:31.067Z DBG httprouter/message.go:65 > response: (200) {"electionId":"98591f8979815b106c84de3299489807e3b3ce39bb6c7c279d97020600000000","organizationId":"5b106c84de3299489807e3b3ce39bb6c7c279d97","status":"READY","startDate":"0001-01-01T00:00:00Z","endDate":"2023-10-04T08:10:21.067294662Z","voteCount":0,"finalResults":false,"manuallyEnded":false,"census":{"censusOrigin":"OFF_CHAIN_TREE_WEIGHTED","censusRoot":"778ee1ac80aef0b79563228364570a3f9a55077c9bd13dcb7f5ddc386bd1461e","postRegisterCensusRoot":"","censusURL":"ipfs://bafybeiezzvgxwv6nykjkbcmqfd55g7smpilvj7nedx2crynyhtp2u2tx5a","maxCensusSize":6},"metadataURL":"ipfs://bafybeialqkqfsyocfh6oi4ngkvvna7xognsfbvfwo57qjbsxlrqxo2zpjm","creationTime":"2023-10-03T10:33:45Z","voteMode":{"serial":false,"anonymous":true,"encryptedVotes":true,"uniqueValues":false,"costFromWeight":false},"electionMode":{"autoStart":true,"interruptible":true,"dynamicCensus":false,"encryptedMetaData":false,"preRegister":false},"tallyMode":{"maxCount":1,"maxValue":1,"maxVoteOverwrites":1,"maxTotalCost":0,"costExponent":10000},"metadata":{"title":{"default":"fdsa"},"version":"1.1","description":{"default":""},"media":{},"meta":{"census":{"type":"token"},"generated":"ui-scaffold","sdk":{"version":"0.3.1"},"token":{"chainID":5,"decimals":18,"defaultStrategy":1,"id":"0xF530280176385AF31177D78BbFD5eA3f6D07488A","name":"Monkeys","size":11,"startBlock":8901659,"status":{"atBlock":9801601,"progress":100,"synced":true},"symbol":"MON","tag":"test2","totalSupply":"82000000000000000000","type":"erc20"}},"questions":[{"choices":[{"title":{"default":"1"},"value":0},{"title":{"default":"1"},"value":1}],"description":{"default":""},"title":{"default":"fdsa"}}],"results":{"aggregation":"discrete-counting","display":"multiple-question"}}}
vocdoni-demo-voconed-1  | 2023-10-03T10:35:31.067Z DBG middleware/logger.go:158 > "GET http://localhost:8081/v2/elections/98591f8979815b106c84de3299489807e3b3ce39bb6c7c279d97020600000000 HTTP/1.1" from 172.21.0.1:57038 - 200 1716B in 435.12µs
vocdoni-demo-voconed-1  | 2023-10-03T10:35:31.103Z DBG censusdb/censusdb.go:134 > loaded tree 778ee1ac80aef0b79563228364570a3f9a55077c9bd13dcb7f5ddc386bd1461e of type ARBO_POSEIDON
vocdoni-demo-voconed-1  | 2023-10-03T10:35:31.103Z DBG httprouter/message.go:65 > response: (200) {"size":11}
vocdoni-demo-voconed-1  | 2023-10-03T10:35:31.103Z DBG middleware/logger.go:158 > "GET http://localhost:8081/v2/censuses/778ee1ac80aef0b79563228364570a3f9a55077c9bd13dcb7f5ddc386bd1461e/size HTTP/1.1" from 172.21.0.1:57038 - 200 12B in 281.408µs
vocdoni-demo-voconed-1  | 2023-10-03T10:35:31.103Z DBG censusdb/censusdb.go:134 > loaded tree 778ee1ac80aef0b79563228364570a3f9a55077c9bd13dcb7f5ddc386bd1461e of type ARBO_POSEIDON
vocdoni-demo-voconed-1  | 2023-10-03T10:35:31.103Z DBG httprouter/message.go:65 > response: (200) {"weight":"82000000000000000000"}
vocdoni-demo-voconed-1  | 2023-10-03T10:35:31.103Z DBG middleware/logger.go:158 > "GET http://localhost:8081/v2/censuses/778ee1ac80aef0b79563228364570a3f9a55077c9bd13dcb7f5ddc386bd1461e/weight HTTP/1.1" from 172.21.0.1:57050 - 200 34B in 251.994µs
vocdoni-demo-voconed-1  | 2023-10-03T10:35:31.103Z DBG censusdb/censusdb.go:134 > loaded tree 778ee1ac80aef0b79563228364570a3f9a55077c9bd13dcb7f5ddc386bd1461e of type ARBO_POSEIDON
vocdoni-demo-voconed-1  | 2023-10-03T10:35:31.103Z DBG httprouter/message.go:65 > response: (200) {"type":"zkweighted"}
vocdoni-demo-voconed-1  | 2023-10-03T10:35:31.103Z DBG middleware/logger.go:158 > "GET http://localhost:8081/v2/censuses/778ee1ac80aef0b79563228364570a3f9a55077c9bd13dcb7f5ddc386bd1461e/type HTTP/1.1" from 172.21.0.1:57038 - 200 22B in 197.144µs
```

**Comment by altergui** [comment]
closing this issue since it was (partially) fixed in #1129 

the census3 <-> voconed communication still suffers issue #1131
